language/proto/generate_test.go

```diff
@@ -188,6 +188,75 @@ func TestGeneratePackage(t *testing.T) {
 	}
 }
 
+// Test generated files that have been consumed by other rules should not be
+// added to the  rule
+func TestConsumedGenFiles(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		// TODO(jayconrod): set up testdata directory on windows before running test
+		if _, err := os.Stat("testdata"); os.IsNotExist(err) {
+			t.Skip("testdata missing on windows due to lack of symbolic links")
+		} else if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	oldContent := []byte(`
+proto_library(
+    name = "existing_gen_proto",
+    srcs = ["gen.proto"],
+)
+proto_library(
+    name = "dead_proto",
+    srcs = ["dead.proto"],
+)
+`)
+	old, err := rule.LoadData("BUILD.bazel", "", oldContent)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	genRule1 := rule.NewRule("proto_library", "gen_proto")
+	genRule1.SetAttr("srcs", []string{"gen.proto"})
+	genRule2 := rule.NewRule("not_proto_library", "gen_not_consumed_proto")
+	genRule2.SetAttr("srcs", []string{"gen_not_consumed.proto"})
+
+	c, lang, _ := testConfig(t, "testdata")
+
+	args := language.GenerateArgs{
+		Config:       c,
+		Dir:          filepath.FromSlash("testdata/protos"),
+		File:         old,
+		Rel:          "protos",
+		RegularFiles: []string{"foo.proto"},
+		GenFiles:     []string{"gen.proto", "gen_not_consumed.proto"},
+		OtherGen:     []*rule.Rule{genRule1, genRule2},
+	}
+
+	args.Config = c
+	res := lang.GenerateRules(args)
+	// Make sure that "gen.proto" is not added to existing foo_proto rule because it is consumed
+	// "gen_not_consumed.proto" is consumed not by "proto_library", this is why it is added
+	if len(res.Gen) != 1 {
```

Comment:
As with `res.Empty` below, create an empty file, insert the rules, format, and compare with a string. That should make it more clear what's expected.